protocols/upnp/src/provider.rs

```diff
@@ -0,0 +1,173 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::{error::Error, fmt, net::Ipv4Addr};
+
+use crate::{
+    behaviour::{GatewayEvent, GatewayRequest},
+    Config,
+};
+use async_trait::async_trait;
+use futures::channel::mpsc::{Receiver, Sender};
+
+/// Protocols available for port mapping.
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq)]
+pub(crate) enum Protocol {
+    Tcp,
+    Udp,
+}
+
+impl fmt::Display for Protocol {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Protocol::Tcp => f.write_str("tcp"),
+            Protocol::Udp => f.write_str("udp"),
+        }
+    }
+}
+
+/// Interface that interacts with the inner gateway by messages,
+/// `GatewayRequest`s and `GatewayEvent`s.
+pub struct Gateway {
+    pub(crate) sender: Sender<GatewayRequest>,
+    pub(crate) receiver: Receiver<GatewayEvent>,
+    pub(crate) addr: Ipv4Addr,
+}
+
+/// Abstraction to allow for compatibility with various async runtimes.
+#[async_trait]
+pub trait Provider {
+    async fn search_gateway(config: Config) -> Result<Gateway, Box<dyn Error>>;
+}
+
+macro_rules! impl_provider {
+    ($impl:ident, $executor: ident, $gateway:ident, $protocol: ident, $options: ident) => {
+        use crate::behaviour::{GatewayRequest, GatewayEvent};
+        use super::{Gateway, Config, Protocol};
+
+        use async_trait::async_trait;
+        use std::error::Error;
+        use futures::{channel::mpsc, select, stream::FuturesUnordered, FutureExt, StreamExt, SinkExt};
+
+        impl From<Protocol> for $protocol {
+            fn from(protocol: Protocol) -> Self {
+                match protocol {
+                    Protocol::Tcp => $protocol::TCP,
+                    Protocol::Udp => $protocol::UDP,
+                }
+            }
+        }
+
+        #[async_trait]
+        impl super::Provider for $impl {
+            async fn search_gateway(config: Config) -> Result<super::Gateway, Box<dyn Error>> {
+                let options = $options {
+                    bind_addr: config.bind_addr,
+                    broadcast_address: config.broadcast_addr,
+                    timeout: config.timeout,
+                };
+                let gateway = $gateway::search_gateway(options).await?;
+                let addr = gateway.get_external_ip().await?;
+
+                let (events_sender, mut task_receiver) = mpsc::channel(5);
+                let (mut task_sender, events_queue) = mpsc::channel(0);
+
+                $executor::spawn(async move {
```

Comment:
I have a slight preference for a single task (i.e. as it is implemented here) vs. one task per request. I am not arguing from a performance perspective, but rather from a complexity perspective. I find a single task handling all IGD requests simpler to reason about than many. This is also in line with https://github.com/libp2p/rust-libp2p/blob/master/docs/coding-guidelines.md#keep-things-sequential-unless-proven-to-be-slow.